packages/snjs/lib/protocol/collection/tag_notes_index.ts

```diff
@@ -0,0 +1,69 @@
+import { ContentType } from '@Models/content_types';
+import { ItemCollection } from './item_collection';
+import { SNNote } from '@Lib/models';
+import { SNTag } from '@Lib/index';
+import { UuidString } from '@Lib/types';
+
+export class TagNotesIndex {
+  private tagToNotesMap: Record<UuidString, Set<UuidString>> = {};
+  private allCountableNotes = new Set<UuidString>();
+
+  constructor(private collection: ItemCollection) {}
+
+  private isNoteCountable = (note: SNNote) => {
+    return !note.archived && !note.trashed;
+  };
+
+  public allCountableNotesCount(): number {
+    return this.allCountableNotes.size;
+  }
+
+  public countableNotesForTag(tag: SNTag): number {
+    return this.tagToNotesMap[tag.uuid]?.size;
+  }
+
+  public receiveTagAndNoteChanges(items: (SNTag | SNNote)[]): void {
+    const notes = items.filter(
+      (item) => item.content_type === ContentType.Note
+    ) as SNNote[];
```

Comment:
With a typeguard you should get the correct type without static casts,

`const isNote = (x: SNItem): x is SNNote => x.content_type === ContentType.Note`

Same for tags, and the call to receiveTagAndNoteChanges in item manager:

```
    // should work and be typesafe.
    const notes: SNote[] = items.filter(isNote );
```

Reply:
Good call 👍